# Assignment1 - Part2 - How many words does BERT know?

In [399]:
# import needed packages
import re

In [400]:
# Load the text file
with open("BERT-vocab.txt", "r") as fp:
  bert_file = fp.read()
  bert_file_list = [line.strip() for line in bert_file.split("\n")]

In [401]:
type(bert_file)

str

In [402]:
len(bert_file_list)

30523

So the total number of words obtained from the Bert file is **30523**

In [403]:
# Remove symbols (unused and some commands like [PAD] [CLS] [SEP])
bert_file_list = [word for word in bert_file_list if not re.compile(r'\[.*\]').match(word)]
len(bert_file_list)

29524

After removing the symbols and commands Bert knows **29524** words 

In [404]:
# Remove words starting with double #
bert_file_list = [word for word in bert_file_list if not re.compile(r'\#\#.*').match(word)]
len(bert_file_list)

23696

There are around 5828 words starting with ## so after removing them Bert knows **23696**

In [405]:
# Remove numbers (if numbers are in front/ middle/ last they are not valid)
bert_file_list = [word for word in bert_file_list if not re.compile(r'.*[0-9]+.*').match(word)]
len(bert_file_list)

22719

Removing numbers Bert knows **22719** words.

In [406]:
# Remove single characters
bert_file_list = [word for word in bert_file_list if not re.compile('^.$').match(word)]
len(bert_file_list)

21732

Single Alphabets are not considered as words. So we remove them and bert knows **21732** words.

In [407]:
# Remove words with starting with special characters
bert_file_list = [word for word in bert_file_list if not re.compile(r'[^a-z]+').match(word)]
len(bert_file_list)

21727

We remove words starting with non-alphabetic characters and bert knows **21727** words.

In [415]:
# Remove most common suffixes
common_suffix_regex = re.compile(r'^[a-z]+(ly|ness|ment|ing|ed|ship|er|hood|lity|ion|le|able|ism|ful)$')
bert_file_list = [word for word in bert_file_list if not common_suffix_regex.match(word)]
len(bert_file_list)

11574

While removing the most common suffixes it is possible that some words in which the above are not suffixes but a part of the word also gets excluded as invalid/unknown words. There are some prefixes like '**al**' which is causing more loss of valid words so have not included them.

In [416]:
# Remove common prefixes (dis, un etc)
common_prefix_regex = re.compile(r'(dis|un|im|non|ir|il|in)[a-z]{4,}')
bert_file_list = [word for word in bert_file_list if not common_prefix_regex.match(word)]
len(bert_file_list)

11574

For prefix when checking we see if the word following has atleast 4 characters minimum. If it's less than 4 we consider them as valid words. For example, **uncle** has **un** as prefix but is still valid. This won't prevent all valid words from getting trimmed but I feel this will reduce the error rate a bit.

In [417]:
# Remove plural forms
bert_file_list = [word for word in bert_file_list if not re.compile(r'^[a-z]{3,}(s|es|ies)$').match(word)]
len(bert_file_list)

11574

When removing plural forms we consider we keep a minimum length of 3 for the preceding string. In this way we can avoid trimming of words such as was, is etc.

# Report:

Bert-Vocabulary has around 30K words from which we do a first round of cleaup like removing single character words, numbers, symbols and non-alphabetic characters. 

After this first level of cleanup, we now eliminate inflected forms of the words. This can be achieved by eliminating words with common prefixes, suffixes and even the plural forms of the word.

I have done these operations using Python regex by eliminating the invalid words in each step and applying the next rule on the filtered list.

Performing the above operations we get a total word count of **11,574**. But in this process we may have lost valid words when dealing with prefix, suffix and plural part as stated in the examples above. 

Also there are few two to three letter words in the result which are invalid but not removed because it would cut out the other meaningful two and three letter words.

Since this assignment is to give an approximate estimate of the number of words known by Bert, we can come to a conclusion that it would in the range of 11,000 to 11,600 words.
